design/external-authorization-design.md

```diff
@@ -0,0 +1,453 @@
+Setting this field without also setting the `TLS` field is an error.
+
+The `AuthorizationServer` field carries and optional default authorization policy.
+This policy sets the default policy for each subsequent `Route`.
+If `authPolicy.Disabled` is true, included `Routes` will not require authorization unless they configure their own authorization policy with `Disabled` set to false.
+The `authPolicy.Context` entries from the root `HTTPProxy` are merged with the context entries on each `Route`.
+Where the two policies have context keys in common, the key from the root `HTTPProxy` is overwritten.
+The merged `Context` entries are send to the authorization service as part of the check request.
+
+Contour sets the following Envoy external authorization configurations that operators cannot configure:
+
+- Authentication services always fail closed
+- TLS client certificates are always sent if they are present
+- The client request body is never sent
+
+## Alternatives Considered
+
+1. Contour could install itself as the authorization server.
+   This could remove some of the limitations of the Envoy configuration structure at the cost of more complexity in Contour.
+1. Integrate external authorization directly into `HTTPProxy`.
+   This increases the complexity of the `HTTPProxy` structure and makes it
+   difficult to re-use the same authorization service acrtoss multiple proxies.
+   A separate CRD gives better opportunities to expose useful operational status.
+
+## Security Considerations
+
+- Most HTTP authorization protocols depend on HTTPS to keep headers and URL query parameters private.
+  Supporting only authorization on HTTPS, Contour reduces the changes of misconfiguring this.
+- HTTPS sessions between Envoy and the authorization server are required.
+  TLS validation can be configured if necessary (should be recommended),
+- Authorization services can run in separate Kubernetes namespaces with limited privilege.
+
+## Compatibility
+
+Contour can support v2 of the Envoy GRPC authorization protocol.
+When Contour moves to v3 of the Envoy API, it ccan support both v2 and v3 authorization protocols.
+
+The `SupportService` CRD supports GRPC authorization support services, but not HTTP ones.
+
+## Implementation
+
+The Contour implementation provides a framework, but isn't useful until authorization services are exist for the protocols that Contour operators need.
+
+The Contour implementation is quite large, but can be separated into smaller stages:
+
+- Support service CRD implementation
+- Support service monitoring and status
+- `HTTPProxy` integration
+- Integration tests
+- Documentation
+
+Since the scope of authorization is quite large, documentation should be written as a new top level section.
+If the documentation is included in the main `httpproxy.md` page, it will be difficult to organize and consume.
+
+### Contour Authserver
```

Comment:
I think we should just re-use existing things instead of building a new one.

Reply:
I agree. However, we need something to test, which is this. If there's an existing Envoy authserver that covers this case, we would support that.